test/integration/test_create_dss_docker_image.py

```diff
@@ -167,3 +168,83 @@ def test_docker_socket_on_host_touched(dss_container_context, fake_docker_socket
         wait_for_socket_access(container)
 
     assert stat_before == socket.stat()
+
+
+class SocketManipulator:
+    def __init__(self, context_provider, socket_on_host: Path):
+        self._context_provider = context_provider
+        self._socket_on_host = socket_on_host
+        self._container = None
+
+    @contextmanager
+    def context(self):
+        with self._context_provider(self._socket_on_host) as container:
+            self._container = container
+            yield self
+            self._container = None
+
+    def run(self, command: str, **kwargs) -> str:
+        """
+        Execute command in container and verify success.
+        """
+        exit_code, output = self._container.exec_run(command, **kwargs)
+        output = output.decode("utf-8").strip()
+        assert exit_code == 0, output
+        return output
+
+    def numeric_gid(self, group_entry: str) -> int:
+        """group_entry is "ubuntu:x:971:", for example"""
+        return int(group_entry.split(':')[2])
+
+    def get_gid(self, group_name: str) -> int:
+        output = self.run(f"getent group {group_name}")
+        return self.numeric_gid(output)
+
+    def get_unassigned_gid(self) -> int:
+        """
+        Return a new gid, that is not used for any other group, yet.
+        """
+        gid = 0
+        for line in self.run("cat /etc/group").splitlines():
+            if not line.startswith(f"nogroup:"):
+                gid = max(gid, self.numeric_gid(line))
+        return gid + 1
+
+    def chgrp_of_docker_socket_on_host(self, gid: str):
+        self.run(f"chgrp {gid} {DOCKER_SOCKET_CONTAINER}", user="root")
+
+    def assert_jupyter_member_of(self, group_name: str):
```

Comment:
Does pytest do a proper assert rewrite if the assert is not in a test case

Reply:
In my opinion, assert rewrite is fine this case.
For demonstration, in class `SocketManipulator` I changed method `get_unassigned_gid()` to provoke an error with
```python
for line in self.run("xcat /etc/group").splitlines():
```

pytest in this case displayed:
```
    def run(self, command: str, **kwargs) -> str:
        # .... [some parts left out]
>       assert exit_code == 0, output
E       AssertionError: OCI runtime exec failed: 
E           exec failed: unable to start container process: exec: "xcat": 
E            executable file not found in $PATH: unknown
E       assert 126 == 0

test/integration/test_create_dss_docker_image.py:192: AssertionError
```

When reading about assert rewrite I have the impression that its more relevant when the `assert` statement is in a different module, which is not the case in this example. 

If you prefer I could also refactor the code to have all `assert` statements in the test itself. 
Please tell me, if I should do that.

Reply:
Please also note an example of a failed test:
https://github.com/exasol/ai-lab/actions/runs/8344507197/job/22837160914#step:7:920

Reply:
Ok, I think, I need to read about rewrite again. Somehow the theory doesn't match with some of my practical experience. But maybe this was in the context of nested functions in tests functions.

Reply:
Would be cool if pytest could log what it is rewriting

Reply:
Yeah, I agree the assert rewrite works